cmd/contour/serve.go

```diff
@@ -216,6 +216,36 @@ func NewServer(log logrus.FieldLogger, ctx *serveContext) (*Server, error) {
 		Scheme:                 scheme,
 		MetricsBindAddress:     "0",
 		HealthProbeBindAddress: "0",
+		NewCache: ctrl_cache.BuilderWithOptions(ctrl_cache.Options{
+			// TransformByObject is a function that allows changing incoming objects before they are cached by the informer.
+			// This is useful for saving memory by removing fields that are not needed by Contour.
+			TransformByObject: ctrl_cache.TransformByObject{
+				&corev1.Secret{}: func(obj interface{}) (interface{}, error) {
+					secret, _ := obj.(*corev1.Secret)
+
+					// Do not touch Secrets that might be needed.
+					if secret.Type == corev1.SecretTypeTLS || secret.Type == corev1.SecretTypeOpaque {
+						return obj, nil
+					}
+
+					// Other types of Secrets will never be referred to, so we can remove all data.
+					// For example Secrets of type helm.sh/release.v1 can be quite large.
+					// Last-applied-configuration annotation might contain a copy of the complete data.
+					secret.Data = nil
+					secret.SetManagedFields(nil)
+					secret.SetAnnotations(nil)
+
+					// Returning error will avoid handlers from being called.
+					// This does not avoid caching.
+					return nil, fmt.Errorf("ignoring secret %s/%s of type %s", secret.Namespace, secret.Name, secret.Type)
+				},
+			},
+			// DefaultTransform is called for objects that do not have a TransformByObject function.
+			DefaultTransform: func(obj interface{}) (interface{}, error) {
+				obj.(client.Object).SetManagedFields(nil)
```

Comment:
I'm thinking we should check whether the type-assertion succeeds, and do _something_ if it doesn't -- probably just return the object as-is. Somewhat related to the above comment, where we might get a `cache.DeletedFinalStateUnknown`.

Reply:
Fixed